File: 12_References_and_Pointers/Linked_List/src/doubly_linked_list.cpp
#include "doubly_linked_list.hpp"

// Node constructor
Node2::Node2(int value) : data(value), next(nullptr), prev(nullptr) {}

// DoublyLinkedList constructor
DoublyLinkedList::DoublyLinkedList() : head(nullptr) {}

// Method to append a node at the end of the list
void DoublyLinkedList::append(int data) {
    Node2* newNode = new Node2(data);
    if (!head) { // If the list is empty
        head = newNode;
        return;
    }
    Node2* last = head;
    while (last->next) { // Traverse to the last node
        last = last->next;
    }
    last->next = newNode; // Link the new node
    newNode->prev = last; // Set the previous pointer of the new node
}

// Method to prepend a node at the beginning of the list
void DoublyLinkedList::prepend(int data) {
    Node2* newNode = new Node2(data);
    if (!head) { // If the list is empty
        head = newNode;
        return;
    }
    newNode->next = head; // Link the new node to the head
    head->prev = newNode; // Set the previous pointer of the head
    head = newNode;       // Update the head to the new node
}

// Method to delete a node by its value
void DoublyLinkedList::deleteNode(int key) {
    Node2* current = head;
    while (current) { // Traverse the list
        if (current->data == key) { // If the node to be deleted is found
            if (current->prev) { // If it's not the first node
                current->prev->next = current->next;
            }
            if (current->next) { // If it's not the last node
                current->next->prev = current->prev;
            }
            if (current == head) { // If it's the head node
                head = current->next;
            }
            delete current; // Free the memory
            return;
        }
        current = current->next;
    }
}

// Method to display the list from head to tail
void DoublyLinkedList::displayForward() {
    Node2* current = head;
    while (current) {
        std::cout << current->data << " <-> ";
        current = current->next;
    }
    std::cout << "nullptr" << std::endl;
}

// Method to display the list from tail to head
void DoublyLinkedList::displayBackward() {
    Node2* current = head;
    if (!current) { // If the list is empty
        std::cout << "nullptr" << std::endl;
        return;
    }
    while (current->next) { // Traverse to the last node
        current = current->next;
    }
    while (current) { // Traverse backward
        std::cout << current->data << " <-> ";
        current = current->prev;
    }
    std::cout << "nullptr" << std::endl;
}

// Destructor to free the memory
DoublyLinkedList::~DoublyLinkedList() {
  Node2* current = head;
  while (current) {
      Node2* nextNode = current->next;
      delete current;
      current = nextNode;
  }
}

File: 12_References_and_Pointers/linked_list/doubly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data  # Store the data
        self.next = None  # Pointer to the next node
        self.prev = None  # Pointer to the previous node

class DoublyLinkedList:
    def __init__(self):
        self.head = None  # Initialize the head of the list

    def append(self, data):
        """Add a node at the end of the list."""
        new_node = Node(data)
        if not self.head:  # If the list is empty
            self.head = new_node
            return
        last = self.head
        while last.next:  # Traverse to the last node
            last = last.next
        last.next = new_node  # Link the new node
        new_node.prev = last  # Set the previous pointer of the new node

    def prepend(self, data):
        """Add a node at the beginning of the list."""
        new_node = Node(data)
        if not self.head:  # If the list is empty
            self.head = new_node
            return
        new_node.next = self.head  # Link the new node to the head
        self.head.prev = new_node  # Set the previous pointer of the head
        self.head = new_node  # Update the head to the new node

    def delete(self, key):
        """Delete a node by its value."""
        current = self.head
        while current:  # Traverse the list
            if current.data == key:  # If the node to be deleted is found
                if current.prev:  # If it's not the first node
                    current.prev.next = current.next
                if current.next:  # If it's not the last node
                    current.next.prev = current.prev
                if current == self.head:  # If it's the head node
                    self.head = current.next
                return
            current = current.next

    def display_forward(self):
        """Display the list from head to tail."""
        current = self.head
        while current:
            print(current.data, end=" <-> ")
            current = current.next
        print("None")

    def display_backward(self):
        """Display the list from tail to head."""
        current = self.head
        if not current:  # If the list is empty
            print("None")
            return
        while current.next:  # Traverse to the last node
            current = current.next
        while current:  # Traverse backward
            print(current.data, end=" <-> ")
            current = current.prev
        print("None")

# Example usage
if __name__ == "__main__":
    dll = DoublyLinkedList()
    dll.append(1)
    dll.append(2)
    dll.append(3)
    dll.prepend(0)

    print("Display forward:")
    dll.display_forward()  # Output: 0 <-> 1 <-> 2 <-> 3 <-> None

    print("Display backward:")
    dll.display_backward()  # Output: 3 <-> 2 <-> 1 <-> 0 <-> None

    dll.delete(2)
    print("After deleting 2:")
    dll.display_forward()  # Output: 0 <-> 1 <-> 3 <-> None
